File: 0001~0500/0028. Implement strStr()/main.cpp
#include<iostream>
#include<string>
using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        int len = haystack.length(), target = needle.length(), i,j;
        if (target == 0)
            return 0;
        if (len == 0)
            return -1;
        for (i=0; i<len-target+1; i++) { //care! reduce time
            for (j=0; j<target; j++) {
                if (haystack[i+j] != needle[j])
                    break;
            }
            if (j == target)
                return i;
        }
        return -1;
    }
};

File: 0001~0500/0028. Implement strStr()/main.py
"""
Source: https://leetcode.com/problems/find-the-index-of-the-first-occurrence-in-a-string/description/
Date: 2023/3/4
Skill: KMP
Ref: shorturl.at/fhyDX
Runtime: 32 ms, faster than 66.84%
Memory Usage: 13.8 MB, less than 54.39%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        suf, dp = [0 for _ in range(len(needle))], [0 for _ in range(len(haystack))]

        def preprocess():
            for i in range(1, len(needle)):
                j = suf[i - 1]
                while j > 0 and needle[i] != needle[j]:
                    j = suf[j - 1]

                suf[i] = j + (needle[i] == needle[j])

        preprocess()
        if needle[0] == haystack[0]:
            dp[0] = 1
        if len(needle) == 1 and dp[0]:
            return 0
        for i in range(1, len(haystack)):
            j = dp[i - 1]
            while j > 0 and haystack[i] != needle[j]:
                j = suf[j - 1]
            dp[i] = j + (haystack[i] == needle[j])
            if dp[i] == len(needle):
                return i - len(needle) + 1

        return -1


if __name__ == "__main__":
    s = Solution()
